packages/server/src/data/db/db-operations/post-ops.js

```diff
@@ -102,7 +104,77 @@ const getPostsByUserID = async (userId, sortType) => {
         ? { score: -1, date: -1, _id: -1 }
         : { date: -1, _id: -1 },
   });
-  return Post.aggregate(aggregation);
+  const posts = await Post.aggregate(aggregation);
+  return posts;
+};
+
+const getVotedPostsByUserID = async (userId, upvote) => {
+  const voteField = upvote ? "upvoters" : "downvoters";
+
+  const aggregation = [];
+  aggregation.push({
+    $match: {
+      [voteField]: { $in: [userId] },
+    },
+  });
+  aggregation.push({
+    $project: {
+      title: true,
+      body: true,
+      date: true,
+      numUpvotes: true,
+      numDownvotes: true,
+      username: true,
+      tag: true,
+      score: { $subtract: ["$numUpvotes", "$numDownvotes"] },
+      isMature: true,
+    },
+  });
+  const posts = await Post.aggregate(aggregation);
+  return posts;
+};
+
+const upvotePost = async (postId, userId) => {
+  const updateSpec = {
+    $push: { upvoters: userId },
+    $pull: { downvoters: userId },
+    $inc: { numUpvotes: 1 },
+  };
+
+  // TODO can avoid making mult async requests if we inferred vote counts from voter array length
+  const post = await Post.findById(postId);
+  if (post.downvoters.includes(userId)) {
+    updateSpec.$inc.numDownvotes = -1;
+  }
+
+  const updatedPost = await Post.findByIdAndUpdate(postId, updateSpec, {
+    new: true,
+  });
+
+  return updatedPost
+    ? { ...updatedPost.toJSON(), comments: undefined } // TODO use $project instead
+    : undefined;
+};
+
+const downvotePost = async (postId, userId) => {
+  const updateSpec = {
+    $push: { downvoters: userId },
+    $pull: { upvoters: userId },
+    $inc: { numDownvotes: 1 },
+  };
+
+  const post = await Post.findById(postId);
+  if (post.upvoters.includes(userId)) {
+    updateSpec.$inc.numUpvotes = -1;
+  }
+
+  const updatedPost = await Post.findByIdAndUpdate(postId, updateSpec, {
```

Comment:
I think we might have to change the where clause to use `$nin` post in downvoters. 

https://docs.mongodb.com/manual/reference/operator/query/nin/

Reply:
Also we should just return the same fields as above for projection, don't need everything:
https://stackoverflow.com/questions/27302398/mongoose-how-to-project-protected-fields-on-update